src/fork.mjs

```diff
@@ -0,0 +1,53 @@
+import ensureIterable from './internal/ensure-iterable'
+import { Exchange } from './internal/queues'
+
+export default function * fork (iterable) {
+  const iterator = ensureIterable(iterable)[Symbol.iterator]()
+
+  let iterableNumber = 0
+  let noNewIterables = false
+  const exchange = new Exchange()
+  let done = false
+
+  function fetch () {
```

Comment:
It is a bit unpleasant that all these function definitions are internal to a particular call of `fork()`, but only perf analysis will tell how much that matters.